### ISSUE TITLE
any upgrade instructions to 0.2.0 and 0.3.0?

### ISSUE THREAD

**Issue by blackpuppy** [JiriChara/vue-kindergarten#19]
I noticed there are new releases 0.2.0 and [0.3.0](https://www.npmjs.com/package/vue-kindergarten) on npmjs.com.  But in this GitHub repo, these 2 new releases are not listed under [Releases](https://github.com/JiriChara/vue-kindergarten/releases).  I tried to upgrade to these new releases using `yarn upgrade`, and met strange errors when running in browser, like:

```
Cannot read property 'isAllowed' of undefined
```

I wrote my code when I was using 0.1.3.  Currently I am using 0.1.4 without a problem.  Maybe I need to make some changes to adapt to 0.3.0?  Is there any instructions or release notes on how to upgrade to these new releases?

Thanks!
Zhu Ming

**Comment by JiriChara** [comment]
Hey @blackpuppy,

I have updated the [Releases](https://github.com/JiriChara/vue-kindergarten/releases) section.

Can you share your code with me? Just the definition of your perimeter and the way you use it in your view.

There shouldn't be any noticeable breaking changes in v0.3.0. All e2e tests has passed and I have updated vue-kindergarten in 2 of my projects and didn't have any issue :thinking:

**Comment by blackpuppy** [comment]
Hi, @JiriChara,

I described my use case below.

I am using Laravel 5.4 with laravel-mix.

```
// resources/assets/js/app.js
import VueKindergarten from 'vue-kindergarten';
import store from './store/';
import child from './child';

Vue.use(VueKindergarten, {
    child,
});

const app = new Vue({
    el: '#app',
    template: '<main-app/>',
    components: { MainApp },
    router,
    store,
});
```

And child is defined as:

```
// resources/assets/js/child.js
export default (store) => {
    return store;
};
```

Sorry, I haven't tried out your fix on child reactivity yet.

The BasePerimeter is using store's getters:

```
// resources/assets/js/perimeters/BasePerimeter.js
import { Perimeter } from 'vue-kindergarten';

export default class BasePerimeter extends Perimeter {
    isAdmin() {
        const allowed = !!this.child
            && this.child.getters['auth/isLoggedIn']
            && this.child.getters['auth/isAdmin'];
        return allowed;
    }

    isAccountManager() {
        const allowed = !!this.child
            && this.child.getters['auth/isLoggedIn']
            && this.child.getters['auth/isAccountManager'];
        return allowed;
    }
}
```

And I have only 1 perimeter:

```
// resources/assets/js/perimeters/userPerimeter.js
import BasePerimeter from './BasePerimeter';

export default new BasePerimeter({
    purpose: 'user',

    govern: {
        'can viewOtherUserTrxns': function () {
            return this.isAdmin() || this.isAccountManager();
        },
    },
});
```

Eventually I use 2 ways to check the permission in Vue components:

```
// resources/assets/js/components/TransactionTable.js
<template>
    <td v-if="$user.isAllowed('viewOtherUserTrxns')"></td>
</template>

<script>
    import userPerimeter from '../perimeters/userPerimeter';

    export default {
        data() {
            if (this.$user.isAllowed('viewOtherUserTrxns')) {
                // ...
            }
            return {
                // ...
            };
        },

        perimeters: [
            userPerimeter,
        ],
    };
</script>
```

All is quite normal except `this.$user.isAllowed(...)` in `data()`.  It works fine with 0.1.4.  Does `userPerimeter` conflict with some intrinsic property in new release?

One more thing I can try is to upgrade my [example](https://github.com/blackpuppy/vue-kindergarten-example) on GitHub.

Thanks!
Zhu Ming

**Comment by JiriChara** [comment]
Hey @blackpuppy,

this one is little bit tricky! There was an issue that components haven't been updated after user logged in: https://github.com/JiriChara/vue-kindergarten/issues/8. The problem was that all exposed methods and purposes like `$user`, `$isAllowed` etc. were not computed properties and therefore not updating.

They are now computed properties and therefore not available in the `data()` method, but you still have 2 options:

```js
data() {
  if (this.$options.computed.$user.isAllowed('viewOtherUserTrxns')) {
    // ...
  }
  // ...
}
```

or

```js
import { createSandbox } from 'vue-kindergarten';

// ...
data() {
  const sandbox = createSandbox(child, {
    perimeters: [
      userPerimeter,
    ]
  });
  if (sandbox.user.isAllowed('viewOtherUserTrxns')) {
    // ...
  }
  // ...
}
```

I would recommend you to upgrade to latest `vue-kindergarten`, because it includes several bugfixes and includes some performance improvements. I will try to find a workaround to make it work again.

**Comment by blackpuppy** [comment]
Hi, @JiriChara,

I have upgraded to 0.3.1 and tried your 2 options.

The first option looks simpler but does not work.  It gives me an warning:

```
[Vue warn]: Error in data(): "TypeError: this.$options.computed.$user.isAllowed is not a function"
```

I need to make a bit tweak to make it work:

```
data() {
  if (this.$options.computed.$user().isAllowed('viewOtherUserTrxns')) {
    // ...
  }
  // ...
}
```

But I am not sure if it is efficient.

The second option is a bit cumbersome as I need to import a few stuff.  But I can make it work.  One thing to change is that since I am using store's getters in child and getters are not available in this new sandbox, I need to avoid using getters here.

I also tried the child reactivity fix and it works as expected.  I will describe it separately in #8.

Thanks!
Zhu Ming
